File: jtms/go/jqueens.go
// N-Queens puzzle solver using dependency-directed search with JTRE.
// Translated from jqueens.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

import (
	"fmt"
	"math"
	"time"
)

// Queens puzzle state
var (
	NAssumptions int
	Placements   [][]interface{}
)

// TestQueens runs the N-Queens solver from `from` to `to`.
// Corresponds to the Lisp function test-queens.
func TestQueens(from, to int) {
	for n := from; n <= to; n++ {
		start := time.Now()
		NQueens(n, false)
		elapsed := time.Since(start)
		fmt.Printf("\n For n=%d, %d solutions, %d assumptions. (%v)",
			n, len(Placements), NAssumptions, elapsed)
	}
}

// NQueens solves the N-Queens puzzle.
// Returns the number of solutions found.
// Corresponds to the Lisp function n-queens.
func NQueens(n int, debugging bool) int {
	jtre := SetupQueensPuzzle(n, debugging)
	choiceSets := MakeQueensChoiceSets(n)
	SolveQueensPuzzle(choiceSets, jtre)
	return len(Placements)
}

// SetupQueensPuzzle initializes the JTRE for the queens puzzle.
// Corresponds to the Lisp function setup-queens-puzzle.
func SetupQueensPuzzle(n int, debugging bool) *Jtre {
	jtre := CreateJtre(fmt.Sprintf("%d-Queens JTRE", n), debugging)
	Placements = nil
	NAssumptions = 0
	RegisterQueensRules(jtre)
	return jtre
}

// MakeQueensChoiceSets creates the choice sets for column placements.
// Each set contains all possible placements for one column.
// Corresponds to the Lisp function make-queens-choice-sets.
func MakeQueensChoiceSets(n int) [][]interface{} {
	choiceSets := make([][]interface{}, 0, n)
	for col := 1; col <= n; col++ {
		columnQueens := make([]interface{}, 0, n)
		for row := 1; row <= n; row++ {
			columnQueens = append(columnQueens, []interface{}{"Queen", col, row})
		}
		choiceSets = append(choiceSets, columnQueens)
	}
	return choiceSets
}

// SolveQueensPuzzle performs dependency-directed search.
// Corresponds to the Lisp function solve-queens-puzzle.
func SolveQueensPuzzle(choiceSets [][]interface{}, jtre *Jtre) {
	if len(choiceSets) == 0 {
		GatherQueensSolution(jtre)
		return
	}
	for _, choice := range choiceSets[0] {
		negChoice := []interface{}{"not", choice}
		if JIn(negChoice, jtre) {
			continue // respect nogood information
		}
		nogood, asns := TryInContext(choice, func() {
			SolveQueensPuzzle(choiceSets[1:], jtre)
		}, jtre)
		NAssumptions++
		if nogood {
			// This assumption lost, justify the negation
			filtered := removeFromSlice(choice, asns)
			justification := append([]interface{}{"Nogood"}, filtered...)
			JAssert(negChoice, justification, jtre)
		}
	}
}

// TryInContext attempts to assume a fact and execute a thunk.
// If a contradiction is found, returns (true, assumptions).
// Otherwise returns (false, nil).
// Corresponds to the Lisp function try-in-context.
func TryInContext(asn interface{}, thunk func(), jtre *Jtre) (bool, []interface{}) {
	tryMarker := []interface{}{"TRY", asn}

	type contradictionResult struct {
		asns []*TmsNode
	}

	var result *contradictionResult

	handler := func(jtmsArg *JTMS, contras []*TmsNode) {
		for _, cnode := range contras {
			asns := AssumptionsOfNode(cnode)
			asnNode := GetJTmsNode(asn, jtre)
			for _, a := range asns {
				if a == asnNode {
					JRetract(asn, tryMarker, true, jtre)
					result = &contradictionResult{asns: asns}
					return
				}
			}
		}
	}

	WithContradictionHandler(jtre.JTMS, handler, func() {
		if JIn(asn, jtre) {
			return
		}
		JAssume(asn, tryMarker, jtre)
		if result != nil {
			return
		}
		RunJRules(jtre)
		if result != nil {
			return
		}
		thunk()
		if result == nil {
			JRetract(asn, tryMarker, true, jtre)
		}
	})

	if result != nil {
		views := make([]interface{}, len(result.asns))
		for i, n := range result.asns {
			views[i] = ViewNode(n)
		}
		return true, views
	}
	return false, nil
}

// QueensOkay checks if two queen positions are non-attacking.
// Corresponds to the Lisp function queens-okay?.
func QueensOkay(x1, y1, x2, y2 int) bool {
	return y1 != y2 && math.Abs(float64(x1-x2)) != math.Abs(float64(y1-y2))
}

// GatherQueensSolution collects a valid solution.
// Corresponds to the Lisp function gather-queens-solution.
func GatherQueensSolution(jtre *Jtre) {
	candidates := JFetch([]interface{}{"Queen", "?c", "?r"}, jtre)
	var solution []interface{}
	for _, q := range candidates {
		if JIn(q, jtre) {
			solution = append(solution, q)
		}
	}
	Placements = append(Placements, solution)
}

// ShowQueensSolution displays a queens solution on the board.
// Corresponds to the Lisp function show-queens-solution.
func ShowQueensSolution(solution []interface{}) {
	n := len(solution)
	for i := 0; i < n; i++ {
		fmt.Println()
		for j := 0; j < n; j++ {
			found := false
			for _, q := range solution {
				qs, ok := q.([]interface{})
				if ok && len(qs) == 3 {
					if deepEqual(qs[1], i) && deepEqual(qs[2], j) {
						found = true
						break
					}
				}
			}
			if found {
				fmt.Print("Q")
			} else {
				fmt.Print("-")
			}
		}
	}
}

// removeFromSlice removes an element from a slice using deepEqual.
func removeFromSlice(elem interface{}, slice []interface{}) []interface{} {
	var result []interface{}
	for _, s := range slice {
		if !deepEqual(s, elem) {
			result = append(result, s)
		}
	}
	return result
}

File: jtms/java/JQueens.java
// N-Queens puzzle solver using dependency-directed search with JTRE.
// Translated from jqueens.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * JQueens implements the N-Queens puzzle using JTRE
 * with dependency-directed backtracking.
 */
public class JQueens {
    private static int nAssumptions = 0;
    private static List<List<Object>> placements = new ArrayList<>();

    /** Run N-Queens from `from` to `to`. Corresponds to test-queens. */
    public static void testQueens(int from, int to) {
        for (int n = from; n <= to; n++) {
            long start = System.currentTimeMillis();
            nQueens(n, false);
            long elapsed = System.currentTimeMillis() - start;
            System.out.printf("%n For n=%d, %d solutions, %d assumptions. (%dms)",
                n, placements.size(), nAssumptions, elapsed);
        }
    }

    /** Solve N-Queens. Returns number of solutions. Corresponds to n-queens. */
    public static int nQueens(int n, boolean debugging) {
        Jtre jtre = setupQueensPuzzle(n, debugging);
        List<List<Object>> choiceSets = makeQueensChoiceSets(n);
        solveQueensPuzzle(choiceSets, jtre);
        return placements.size();
    }

    /** Setup the puzzle. Corresponds to setup-queens-puzzle. */
    public static Jtre setupQueensPuzzle(int n, boolean debugging) {
        Jtre jtre = new Jtre(String.format("%d-Queens JTRE", n), debugging);
        placements = new ArrayList<>();
        nAssumptions = 0;
        JQRule.registerQueensRules(jtre);
        return jtre;
    }

    /** Create choice sets. Corresponds to make-queens-choice-sets. */
    public static List<List<Object>> makeQueensChoiceSets(int n) {
        List<List<Object>> choiceSets = new ArrayList<>();
        for (int col = 1; col <= n; col++) {
            List<Object> columnQueens = new ArrayList<>();
            for (int row = 1; row <= n; row++) {
                columnQueens.add(Arrays.asList("Queen", col, row));
            }
            choiceSets.add(columnQueens);
        }
        return choiceSets;
    }

    /** Solve using dependency-directed search. Corresponds to solve-queens-puzzle. */
    public static void solveQueensPuzzle(List<List<Object>> choiceSets, Jtre jtre) {
        if (choiceSets.isEmpty()) {
            gatherQueensSolution(jtre);
            return;
        }
        for (Object choice : choiceSets.get(0)) {
            List<Object> negChoice = Arrays.asList("not", choice);
            if (JData.isIn(negChoice, jtre)) continue;

            boolean[] nogood = {false};
            List<Object>[] asns = new List[]{null};

            tryInContext(choice, () -> {
                solveQueensPuzzle(choiceSets.subList(1, choiceSets.size()), jtre);
            }, jtre, nogood, asns);

            nAssumptions++;
            if (nogood[0] && asns[0] != null) {
                List<Object> filtered = asns[0].stream()
                    .filter(a -> !JData.deepEqual(a, choice))
                    .collect(Collectors.toList());
                List<Object> justification = new ArrayList<>();
                justification.add("Nogood");
                justification.addAll(filtered);
                JData.jassert(negChoice, justification, jtre);
            }
        }
    }

    /** Try assumption in context with contradiction handler. */
    private static void tryInContext(Object asn, Runnable thunk, Jtre jtre,
                                      boolean[] nogoodOut, List<Object>[] asnsOut) {
        Object tryMarker = Arrays.asList("TRY", asn);
        final boolean[] done = {false};

        jtre.getJtms().withContradictionHandler((jtmsArg, contras) -> {
            for (JTMS.TmsNode cnode : contras) {
                List<JTMS.TmsNode> assumptions = jtmsArg.assumptionsOfNode(cnode);
                JTMS.TmsNode asnNode = JData.getTmsNode(asn, jtre);
                if (assumptions.contains(asnNode)) {
                    JData.jretract(asn, tryMarker, true, jtre);
                    nogoodOut[0] = true;
                    asnsOut[0] = assumptions.stream()
                        .map(n -> (Object) JData.viewNode(n))
                        .collect(Collectors.toList());
                    done[0] = true;
                    return;
                }
            }
        }, () -> {
            if (JData.isIn(asn, jtre)) return;
            JData.jassume(asn, tryMarker, jtre);
            if (done[0]) return;
            JRules.runRules(jtre);
            if (done[0]) return;
            thunk.run();
            if (!done[0]) {
                JData.jretract(asn, tryMarker, true, jtre);
            }
        });
    }

    /** Check if two queens are non-attacking. Corresponds to queens-okay?. */
    public static boolean queensOkay(int x1, int y1, int x2, int y2) {
        return y1 != y2 && Math.abs(x1 - x2) != Math.abs(y1 - y2);
    }

    /** Gather a solution. Corresponds to gather-queens-solution. */
    private static void gatherQueensSolution(Jtre jtre) {
        // Collect all Queen facts that are IN
        List<Object> solution = new ArrayList<>();
        Dbclass queenClass = JData.getDbclass(Arrays.asList("Queen"), jtre);
        for (Datum d : queenClass.getFacts()) {
            if (JTMS.inNode(d.getTmsNode())) {
                solution.add(d.getLispForm());
            }
        }
        placements.add(solution);
    }

    /** Display a solution. Corresponds to show-queens-solution. */
    @SuppressWarnings("unchecked")
    public static void showQueensSolution(List<Object> solution) {
        int n = solution.size();
        for (int i = 0; i < n; i++) {
            System.out.println();
            for (int j = 0; j < n; j++) {
                boolean found = false;
                for (Object q : solution) {
                    if (q instanceof List) {
                        List<Object> ql = (List<Object>) q;
                        if (ql.size() == 3 && JData.deepEqual(ql.get(1), i)
                            && JData.deepEqual(ql.get(2), j)) {
                            found = true;
                            break;
                        }
                    }
                }
                System.out.print(found ? "Q" : "-");
            }
        }
    }

    public static int getAssumptions() { return nAssumptions; }
    public static List<List<Object>> getPlacements() { return placements; }
}
